xbmc/utils/StringUtils.cpp

```diff
@@ -327,96 +203,422 @@ std::wstring StringUtils::FormatV(const wchar_t *fmt, va_list args)
+  return utf8Str;
+}
+
+std::u32string StringUtils::ToUtf32(const std::string_view str)
+{
+  std::u32string utf32Str =
+      CUnicodeConverter::Utf8ToUtf32(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return utf32Str;
+}
+
+std::u32string StringUtils::ToUtf32(const std::wstring_view str)
+{
+  std::u32string utf32Str;
+  utf32Str = CUnicodeConverter::WToUtf32(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return utf32Str;
+}
+
+std::wstring StringUtils::ToWstring(const std::string_view str)
+{
+  std::wstring wStr = CUnicodeConverter::Utf8ToW(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return wStr;
+}
+
+std::wstring StringUtils::ToWstring(const std::u32string_view str)
+{
+  std::wstring result;
+  result = CUnicodeConverter::Utf32ToW(str, FAIL_ON_ERROR, SUBSTITUTE_BAD_CHARS);
+  return result;
+}
+
+namespace
+{
+static int FoldAndCompare(const std::u32string_view str1, const std::u32string_view str2)
+{
+  // FoldAndEquals is faster for equality checks.
+
+  size_t length = std::min(str1.length(), str2.length());
+  for (size_t i = 0; i < length; i++)
+  {
+    char32_t fold_c1 = FoldCaseChar(str1[i]);
+    char32_t fold_c2 = FoldCaseChar(str2[i]);
+
+    if (fold_c1 != fold_c2)
+    {
+      if (fold_c1 < fold_c2)
+        return -1;
+      else
+        return 1;
+    }
+  }
+  // Break a tie using length
+
+  if (str1.length() < str2.length())
     return -1;
-  else if (*(const wchar_t*)a >  *(const wchar_t*)b)
+  if (str2.length() < str1.length())
     return 1;
+
   return 0;
 }
+} // namespace
 
-wchar_t tolowerUnicode(const wchar_t& c)
+std::locale StringUtils::GetCLocale()
 {
-  wchar_t* p = (wchar_t*) bsearch (&c, unicode_uppers, sizeof(unicode_uppers) / sizeof(wchar_t), sizeof(wchar_t), compareWchar);
-  if (p)
-    return *(unicode_lowers + (p - unicode_uppers));
+  static std::locale C_LOCALE;
+  static bool initialized{false};
 
-  return c;
+  if (!initialized)
+  {
+    std::locale x;
+    try
+    {
+      x = std::locale("C.UTF-8"); // MacOS does not like C.UTF-8
+      C_LOCALE = x;
+      initialized = true;
+    }
+    catch (std::runtime_error& e)
+    {
+      // Have to be careful about logging in StringUtils. Can get recursive during
+      // formatting of message.
+
+      std::cout << "Locale C.UTF-8 not supported trying 'C'" << std::endl;
+      try
+      {
+        x = std::locale("C"); // This shouldn't fail
+        C_LOCALE = x;
+        initialized = true;
+      }
+      catch (std::runtime_error& e2)
+      {
+        // Have to be careful about logging in StringUtils. Can get recursive
+
+        std::cout << "Locale C not supported" << std::endl;
+      }
+    }
+  }
+  return C_LOCALE;
 }
 
-wchar_t toupperUnicode(const wchar_t& c)
+std::wstring ToUpperC(const std::wstring_view str)
 {
-  wchar_t* p = (wchar_t*) bsearch (&c, unicode_lowers, sizeof(unicode_lowers) / sizeof(wchar_t), sizeof(wchar_t), compareWchar);
-  if (p)
-    return *(unicode_uppers + (p - unicode_lowers));
+  std::wstring result;
```

Comment:
I think that this should be done selectively.

For all but Windows the stack-based output buffer is sized to contain the same number of code-units (characters) as input + 512 pad bytes. This should normally be more than enough for the conversion, so the conversion loop should normally be one iteration. The output string will be appended to once.

For DARWIN, there is a wrinkle. DARWIN is set to use UTF-8-MAC, which causes iconv to convert and normalize the string. The particular normalization used only expands the string. The amount expands depends upon the characters. Languages like Korean tend to expand more than Latin script based languages. 

For lack of any better guidance, I'll double the stack buffer size for Darwin when encoding
is UTF-8.

Windows c++ compiler forces us to use a fixed-size output buffer which is set to be enough for a 2K u32string. Fortunately I can pre-allocate the output string to be the correct size.

Reply:
Yeah please apply it where it makes sense (like here). It might be a better idea to use alloca or another solution (e.g., use fixed stack buffer and fall back to heap if the string is too big) instead of dynamic-sized arrays on the stack since this is not a C++ standard feature.

Reply:
I like using memory from stack, but there are warnings about blowing the stack. Should at least add a size limit. It is non-standard but g++ and clang both take it. Using alloc or _alloc (Windows) begins to sound more complicated.

It looks like I could use the result string directly as a buffer- std::string.data(). 

- use reserve to preallocate for expected size, which should be exact (except for Darwin & UTF-8-MAC which can produce longer strings that input, but not shorter). 
- use reserve to grow, if needed
- perhaps use resize_to_fit when done

Another option is thread-local malloc buffers, of limited size, but must worry about leaks.
 
I think either stack (with size limit) or using the result string as the buffer is the way to go